### ISSUE TITLE
Tacx Fortius Disconnects

### ISSUE THREAD

**Issue by Mickyduck55** [GoldenCheetah/GoldenCheetah#3812]
During testing of recent Artifacts which are addressing issues with the Fortius not working in slope mode I have come across an issue where the trainer disconnects from GC.  I am working my way back through builds so I can try to find the point at which this broke.
I am using Windows 10 on an i7 HP EliteBook
GC is in train mode using video chart.
I had the issue using The Argus tour initially but for ease of repeatability I am now using Flat course (TACX Amsterdam city trip).  
ci 3627 disconnects almost immediately
ci 3625 disconnect at 17 mins and 10 mins so varies
I will add to this as I work my way backwards.

At the suggestion of @mattipee  I jumped back to ci3604.. same scenario as above 31 mins NO PROBLEMS

**Comment by mattipee** [comment]
Thanks @Mickyduck55 

@michelrdagenais , tagged you for second opinion, please.

Firstly, here is a list of the relevant builds:

![image](https://user-images.githubusercontent.com/15227658/106389375-dc00ea80-63da-11eb-9784-702ad7183186.png)

The top one is the current PR which Micky says fails immediately - I can't see in the code anything that would cause it. A second opinion there would be appreciated.

My hunch says if we have introduced an issue causing instability over time, then it's likely one of:
- ci.3609 usb_set_altinterface()
- ci.3610 moving the msleep(50)
- ci.3620 request/reply changes

The fact that ci.3604 appears to test ok, supports this.

Unfortunately, the time-to-failure for Micky seems long... at least it makes repeat testing a physical process.

Micky - if ci.3609 fails for you, that could support ci.3609 "altinterface" being an issue - I'll prepare a branch to reintroduce the call on Windows only, and can turn that into a PR if your testing supports that.

Otherwise, thank you for all the testing, particularly the effort in fetching and reinstalling builds to diagnose.

**Comment by Mickyduck55** [comment]
Thanks for the suggestions @mattipee and @michelrdagenais, I will get on ci 3609 tomorrow

**Comment by mattipee** [comment]
The other key bit of information here is that the artefact produced from #3725 (I think) appears to work without disconnecting.

**Comment by Mickyduck55** [comment]
That is correct as far as I can tell having used this Artefact frequently
in the last few weeks.

Before yesterday I had two disconnects one was after over an hour and @Michel
Dagenais <michelrdagenais@gmail.com> suggested that it may be static
buildup and described a solution by grounding the trainer, I have not done
this.  The other was in recent testing of rolling calibration which
occurred after 10 mins.

Going forward I will be more specific re which "ci" I am using

**Comment by amtriathlon** [comment]
IMO the first suspect is https://github.com/GoldenCheetah/GoldenCheetah/commit/97ce2989bf734491488222e94968b12c802c7fc9, I would partially revert it, restoring the removed code and changing the condition from OperatingSystem != OSX to OperatingSystem == WINDOWS

**Comment by michelrdagenais** [comment]
For the calibration patches, I imagine that Micky could easily refrain from calibrating and see if there is still a problem. CRS altutide should not be a problem. The various patches that change the force computation, including the WindResistance, could "sensitize" the Fortius differently and make the connection more fragile but it is not too likely. The timing patches, request/reply and msleep, change the behavior of the communication with the Fortius more, but this is not necessarily different from one OS to the other.

I agree that usb_set_altinterface is something to look at closely since I exclusively use Linux and I believe that Mattipee is also on Linux. It could be interesting to add a qDebug there to see if that code (using the alt interface) is called frequently. That would confirm that removing it  would be problematic. It would also be useful to have Micky report his Fortius model version, now that this information is available.

**Comment by mattipee** [comment]
@michelrdagenais thank you.

I will open a PR with the change Ale suggests above, so that the artefact builds for Mick.

Mick has confirmed "good" (or least "not bad, yet") with ci.3604. I think getting a datapoint from him on ci.3609 would be good, particularly if it fails. And then again with the PR.

I am on Debian 10.7. I was apparently able to command the trainer (change resistance), but unable to receive any responses from it, without first disabling the call to usb_set_altinterface. Interested to know whether you found this, I'm assuming though perhaps it was working fine for you with the call, and I'm the "outlier".

**Comment by Mickyduck55** [comment]
@mattipee @michelrdagenais 
FYI
Brake firmware 4100
Serial 5997
Year 2008
Type T1941
Version2=2327
Motobrake=1

**Comment by Mickyduck55** [comment]
@Michel Dagenais <michelrdagenais@gmail.com> re calibration .. I had not
calibrated when these disconnects happened. I found with TTS that provided
I pump the tyre up to the same PSI (120) the calibration does not change so
it is not something I do very frequently especially with how cumbersome it
is in TTS and was in GC until @mattipee <tunabiscuits796@gmail.com> added
rolling calibration.

**Comment by mattipee** [comment]
Micky... ci.3630 for PR #3813, where usb_set_altinterface has been reinstated on Windows.

**Comment by Mickyduck55** [comment]
I have downloaded 3632... but I will skip back to 3630.....

**Comment by mattipee** [comment]
> I have downloaded 3632... but I will skip back to 3630.....

Yes please...
- ci.3630 is the one that reinstates usb_set_altinterface for Windows
- ci.3632 is a result of a commit to the "intensity" PR and not the one you want

With luck, ci.3609 will fail for you, and ci.3630 will pass. Over to you... :)

**Comment by Mickyduck55** [comment]
Yes still getting my head round how GitHub / PR stuff... again I assumed
incorrectly that the updates were cumulative....

**Comment by mattipee** [comment]
Micky reported, while launching GC on console with --debug

> 19 mins to fail 3630
> usb read error  -116
> Started at loss of connection window was blank except for the header.. had this issue a long time ago Ale told me its a USB read error... That was before Michael did some fixing....

116 stale file handle?

**Comment by michelrdagenais** [comment]
The fix that I contributed was https://github.com/michelrdagenais/GoldenCheetah/tree/FortiusConnectError. It simply emits a notification, if the Fortius Run thread does not respond, instead of showing a Message Box. It avoids being stuck in message boxes but does not change the logic.

Currently, the Fortius driver simply quits on USB errors. Errors when reading are shown on qDebug and errors on opening the port or writing result in a "quit" (exit). If Windows has errors from time to time, the fact that we are reading twice as much now (always skipping a packet when reading) can impact the frequency / visibility of the problem. It would not be difficult to try some recovery strategy (close, reopen the port, retry) to be more robust.

Do we have other Windows Fortius users? Is there a specific problem with Micky's setup: cable too long, too many devices draining power from USB, flaky USB port or cable...?

**Comment by mattipee** [comment]
@michelrdagenais Despite three attempts and some forum scraping, I've been unable to run .exe on my Windows7, due to reported missing dll. I've tried fresh install of 3.5, 3.6 snapshot and CI builds, and various combinations of vcredist removal/install/reboot. If I can get it working, I'd be a good candidate for Windows testing. :)

@MickyDuck55 ... please could you also test 3609? This removed usb_set_altinterface but precedes the timing/number-of-read changes.

**Comment by amtriathlon** [comment]
> Despite three attempts and some forum scraping, I've been unable to run .exe on my Windows7, due to reported missing dll.

See https://github.com/GoldenCheetah/GoldenCheetah/wiki/ANT--and-BTLE-Troubleshooting#btle-on-windows

**Comment by Mickyduck55** [comment]
3609 no issues 31 mins

**Comment by Mickyduck55** [comment]
USB Cable is 1,5 m and 1m long I have tried 2 different ones
Other USB connections 1 ANT+ stick and a remote mouse dongle
I know of two other Windows users BUT I believe they are running the
3779... That was the link I sent them a few weeks ago.
I have another PC I can try this on?
Cheers

**Comment by mattipee** [comment]
> 3609 no issues 31 mins

If you had the appetite, 3610 (msleep move) then 3620 (double reads) may help pin it down.

**Comment by Mickyduck55** [comment]
Not today Matt.. but I will be hungry tomorrow....

**Comment by mattipee** [comment]
> > Despite three attempts and some forum scraping, I've been unable to run .exe on my Windows7, due to reported missing dll.
> 
> See https://github.com/GoldenCheetah/GoldenCheetah/wiki/ANT--and-BTLE-Troubleshooting#btle-on-windows

Thanks, Ale.

> A workaround is to copy Qt5Bluetooth.dll from version 3.5, that would allow GC to run without BTLE support.

Unfortunately, 3.5 doesn't launch either, with a different DLL error. I will try again, harder, before seeking further help on -users.

**Comment by mattipee** [comment]
> Not today Matt.. but I will be hungry tomorrow....

I wasn't suggesting today. :)

I feel a bit lazy simply asking for subsequent build tests without producing some targetted debug on a PR to help you out, but I'm short on time myself. The datapoints of those two builds would nevertheless be helpful though, I think.

I'll try to get myself running on Windows.

**Comment by Mickyduck55** [comment]
Its not a problem.. as I say I am delighted to be somewhere with this.
Cheers

**Comment by amtriathlon** [comment]
> Unfortunately, 3.5 doesn't launch either, with a different DLL error. I will try again, harder, before seeking further help on -users.

From 3.5 you only need Qt5Bluetooth.dll to replace the one shipped with 3.6, no need to run, to install v3.6 the first time follow the steps at the end of https://github.com/GoldenCheetah/GoldenCheetah/releases/tag/v3.6-DEV2101, but replacing Qt5Bluetooth.dll

**Comment by mattipee** [comment]
ci.3627 tested for 1hour, no issues on Win7, (20km, -1.2%, spoke magent to cadence sensor)

**Comment by Mickyduck55** [comment]
If I get an issue today I will try the same build on a different PC with
different USB connection

**Comment by Mickyduck55** [comment]
3627 was the most unstable for me so I will test that first as it broke
frequently and within a very short period of time.  If it does break I will
try it on another PC.
@mattipee <tunabiscuits796@gmail.com> what is the significance of the
"spoke magnet" I am using the TACX wired sensor

**Comment by mattipee** [comment]
> 3627 was the most unstable for me so I will test that first as it broke frequently and within a very short period of time. If it does break I will try it on another PC.

Michel's PR has built as ci.3636 with the additional debug, but ci,3627 doesn't have that. However a test of either on another computer might be interesting.

Debug from ci.3636 on disconnect should be the priority though, I think.


> what is the significance of the "spoke magnet" I am using the TACX wired sensor

The Fortius stops if it detects no cadence, the spoke magnet fakes cadence to the wired sensor, so on a downhill section, it will run unattended.

**Comment by Mickyduck55** [comment]
I have downloaded both of them.. just interested in 3627 as it broke almost
immediately and frequently.  The spoke magnet must really mess up the
cadence;).. I know that is the behaviour in TTS I thought it did freewheel
in GC, I commented on one of the earlier builds that it was nice that I
could free wheel.  Maybe I didnt give it long enough to stop. I will test
that today as well.
Thanks

**Comment by mattipee** [comment]
> The spoke magnet must really mess up the cadence;).. I know that is the behaviour in TTS I thought it did freewheel in GC, I commented on one of the earlier builds that it was nice that I could free wheel. Maybe I didnt give it long enough to stop. I will test that today as well. Thanks

At 20kph, the cadence gets to 170 or something... the magnet on a spoke means wheel revolution is used in place of pedal revolution and will make the controller believe you're still pedalling, ... it will only freewheel "while you are pedalling" otherwise the motor stops.

This means I can freewheel an hour's test... 20km at -1.2%.

**Comment by mattipee** [comment]
> At 20kph, the cadence gets to 170 or something... the magnet on a spoke means wheel revolution is used in place of pedal revolution and will make the controller believe you're still pedalling, ... it will only freewheel "while you are pedalling" otherwise the motor stops.
> 
> This means I can freewheel an hour's test... 20km at -1.2%.

@michelrdagenais do you think it's possible we're somehow populating a field in a sent command "out-of-spec" or something, that causes the issues? Something spurious. I don't know how tolerant the head unit would be to that.

Thinking about my "spoke magnet" test above which is already fake in the sense that the resistance values back and forth are constrained to a very narrow range... USB in terms of message volume and timing is already being exercised having started a workout but not even pedalling - at rest. @Mickyduck55 - you could try your favourite-failing-build, start a workout but just leave it alone and see if it fails.

I know there's no real yes/no answer here, just scratching my head... as there may be value in dumping all the sent commands and seeing if there's anything odd at the point of failure that has occurred. May well not be the issue.

**Comment by Mickyduck55** [comment]
@michelrdagenais @mattipee 
So today loaded 3627 with SA Argus film, this is a rather high def film so wanted to exclude video issue.  After about 10 mins "cannot connect to Fotius" comes up and the video window is black.  This is a little different to before usually the connection is lost and video continues to play.
Tried reconnecting which it did after 4 attempts.
Loaded Amsterdam film which I have used for all recent testing as its flat.. about 6 mins in connection lost but video continued to play.. so 3627 does not work for me.
Opened 3636 in debug mode selected Amsterdam pressed connect the message asking if brake was switched on displayed. Physically unplugged USB reinserted and brake found cycled 40 mins without no issues as the user but looks like there are issues in the log
Hope this helps
[michael.txt](https://github.com/GoldenCheetah/GoldenCheetah/files/5911379/michael.txt)

**Comment by michelrdagenais** [comment]
@Mickyduck55 @mattipee 
It looks like the messages in michael.txt are all related to the device not working initially. After a few retries, the thread quits. After your reconnection, the brake motor version is found properly and there are no further error messages.

There seems to be something special in the simspeed branch, perhaps some "out-of-spec" values in the commands as you suggest, since the timing and protocol appear unchanged. There could be an error in the slope calculation, but that would be felt by the rider anyway?

Here is a possible avenue for diagnosis. The debug code should not alter the behavior until an error is detected, at which point the retries may hide the problem but a message will be available. I could improve a bit the debug code (add a timestamp and log the offending bytes for write errors). Then, I could merge the simspeed branch on the debug code. Micky could test that and it should provide a clear trace of the problems he is encountering. What do you think?

**Comment by mattipee** [comment]
Timestamp feels useful. Thing with possible out-of-spec values is that it would likely be the send before a fault, rather than bytes upon write errors.

**Comment by michelrdagenais** [comment]
Currently, the code just quits upon a write error in sendRunCommand. With the debugging code, we will see the error code and if the retries solve the problem. If there is a glitch every 10 or 20 minutes, solved by a retry, it may not be that bad. We will see the error code, where it happens and the offending command. The previous commands may or not be helpful. If we feel that we need more context after the first debug runs, we could add a buffer of the last 10 write commands, and print that out as context when an error is found.

**Comment by Mickyduck55** [comment]
I am happy to test whatever you put together.. WRT  "error in the slope
calculation, but that would be felt by the rider anyway"  Im cycling a flat
course with a coupe of 2% short slopes so my cadence is steady and I do not
feel any anomalous resistance

**Comment by Mickyduck55** [comment]
@michelrdagenais @mattipee 
So far today I have cycled 30 + mins on 3638 and 3639 without any issues.... I have downloaded 3639 again just in case I have messed up which version I am actually using.  Later I will test this again and 3637
I guess the log files limited in size as when I start there is quite a bit of header type information :brake and route details.. this information is  not in the files when I finish?
[3638.txt](https://github.com/GoldenCheetah/GoldenCheetah/files/5925265/3638.txt)
[3639.txt](https://github.com/GoldenCheetah/GoldenCheetah/files/5925268/3639.txt)

**Comment by michelrdagenais** [comment]
The files should be very large, about 60 lines per second. If you are using some console logging option, it may simply keep the last 5000 lines or so. A better way is to start GC from the command line and use file redirection to capture everything in a file. On Linux it is something like "src/GoldenCheetah 2> outfile.txt". A similar mechanism is available on Windows.

Normally, 3639 has the exact same behavior as 3637, except for more debugging info. Testing again 3637 in the same session will indeed be useful to confirm if the problem is in the software version or due to something else like humidity / static electricity. The fact that you got the same result on a different PC would point to a problem in the trainer, if the problem is not in the software version after all.

**Comment by Mickyduck55** [comment]
H Michael
To reassure myself that I am not crazy I have just run 3637 this afternoon
blew up after about 6 mins with the -116 error
Exited GC, unplugged USB started 3639 no problems after 16 min.

**Comment by michelrdagenais** [comment]
The same simspeed code is present in both. Only the debugging code was changed. That debugging code was not present in ci.3627 which also had problems. It looks like a corruption problem that gets sensitized by the code layout, which changes with every code addition. I don't see anything suspicious that would cause corruption in the simspeed code. Ideas anyone @mattipee ? At this point we could make small changes, try to use usb_bulk_read instead of usb_interrupt_read, but your current results don't really point us in a specific direction.

**Comment by Mickyduck55** [comment]
Yes I know its not much help.... as above I will use 3639 again tomorrow with the debug writing to a file as you suggested and just see what happens.  I understand If I cant break it you cannot fix it .....

**Comment by mattipee** [comment]
I'm currently stumped for best direction from here. If I had a spare head unit, I'd swap with you, @Mickyduck55 . Not certian even that would help.

I've always been confused why the intensity PR made any difference. The fact that sometimes an unplug/replug is required doesn't bode well for a software fix, unless there's some kind of "hard reset" we can do... I'm not too familiar with USB.

**Comment by Mickyduck55** [comment]
@mattipee I will try Susan's head unit tomorrow and see if the behaviour is any different.

**Comment by michelrdagenais** [comment]
> I've always been confused why the intensity PR made any difference. 

With the most recent results, it appears to be somewhat independant.

> The fact that sometimes an unplug/replug is required doesn't bode well for a software fix, unless there's some kind of "hard reset" we can do... I'm not too familiar with USB.

It should be possible to reset a USB device from software, at least on Linux. It may cause some delays though which may not be pleasant if it happens every few minutes. I could integrate my debugging code into the master branch with errors always output with qDebug, and the rest of the communication output on qDebug based on some configuration flag. Then, when Micky gets a new version that disconnects, the debugging code will be there. If he never gets a new version that disconnects... the mystery will remain but he will be able to train.

**Comment by Mickyduck55** [comment]
@michelrdagenais @mattipee ... like you both I have no idea why the 3637 seems to be very fragile and 3639 is not, maybe there is a corruption on my install.. I can get a new copy and try that as well

**Comment by michelrdagenais** [comment]
@Mickyduck55 since you were getting the same result on another computer, what is different and what is common? Do you have the exact same version of Windows and drivers, the same file downloaded for GC?

**Comment by Mickyduck55** [comment]
As far as I know they are both the same although different manufacturers. Windows 10 both updated yesterday, drivers installed from the same file copied from a link on FortiusANT.  Same version of GC from the artefacts and the same behaviour, both do not run 3637 for  very long.  TACX Hardware is the same although I did try different USB cable as mentioned above.

Same athlete folder and config copied to the other laptop... 

I need to test 3639 again tomorrow.. as I said almost 50 mins on it today without issue.

I know you say 3639 and 3627 are the same except for the debugging but the behaviour for me is different.. Maybe the debugging is slowing something down in the transfer of data so it doesn't disconnect, I really have no idea but if 3639 does not error then we are good?  TBH I guess you would expect the debug to hinder rather than help as yet another process is layered onto the transfer

It does not look like writing a file works with the -debug switch, I think this is the windows syntax

goldencheetah.exe - debug >c:\GC RLV\OUTPUT.txt

**Comment by michelrdagenais** [comment]
> I know you say 3639 and 3627 are the same except for the debugging but the behaviour for me is different.. Maybe the debugging is slowing something down in the transfer of data so it doesn't disconnect, I really have no idea but if 3639 does not error then we are good? 

You can try 3639 without the -debug option, to see if the debug output makes a difference. However, you have a good point, when debugging is redirected to a file, it should be pretty efficient and have little impact. The debugging output gets buffered and is written to disk in large batches, in parallel, asynchronously. However, if the debug output is sent to the console (and then saved from the console), this is much less efficient and can significantly affect the timing. Each debug line is sent to the console right away, the console uses the graphics libraries to draw pixels for each character with a nice font...

> It does not look like writing a file works with the -debug switch, I think this is the windows syntax
> 
> goldencheetah.exe - debug >c:\GC RLV\OUTPUT.txt

By default, > is like 1>, redirect the channel 1, standard output. What you want is 2>, redirect channel 2, standard error. It should work on Windows. It was inspired from Unix which predates DOS... This should remove the timing disruption effect and should be a much better test.

**Comment by Mickyduck55** [comment]
@michelrdagenais @mattipee 
Well gents I really am at a loss....
Today 
Changed head unit and used different USB Port

3637 lost connection almost immediately
3639 No problems 15 mins
3637 Lost connection after 4 mins
3639 No Problem 15 mins
NOW FOR THE KICKER 3637 no problems 15 mins... 
3639 No problems 5 min including rolling Cal.

I cannot get the debug written to file... I think his is the correct syntax which should do it.. the file is created but its blank and the consul is still displaying the data 
START "" "C:\program files\3639\golden cheetah\goldencheetah.exe" - debug >C:\OUTPUT.TXT 2>&1

So I guess we will just have to move on from 3639?  I tried rolling cal on my last session.. lovely and I was faster and more powerful after ;)

**Comment by michelrdagenais** [comment]
> 3637 lost connection almost immediately
> 3639 No problems 15 mins
> 3637 Lost connection after 4 mins
> 3639 No Problem 15 mins
> NOW FOR THE KICKER 3637 no problems 15 mins...
> 3639 No problems 5 min including rolling Cal.

Something about the unit warming up or simple luck?

> I cannot get the debug written to file... I think his is the correct syntax which should do it.. the file is created but its blank and the consul is still displaying the data
> START "" "C:\program files\3639\golden cheetah\goldencheetah.exe" - debug >C:\OUTPUT.TXT 2>&1

Is this executed in a command shell that will do the redirection, the "START" would not be needed in a shell? The syntax otherwise appears correct but you could try 2> (instead of > and 2>&1).

> So I guess we will just have to move on from 3639? I tried rolling cal on my last session.. lovely and I was faster and more powerful after ;)

The file redirection would really make a difference. Yes, calibration is nice. Perhaps you are getting faster only because of all this test/training :-)

**Comment by Mickyduck55** [comment]
> Something about the unit warming up or simple luck?
Pure luck... it was pretty warm after 15 mins..

> Is this executed in a command shell that will do the redirection, the "START" would not be needed in a shell? The syntax otherwise appears correct but you could try 2> (instead of > and 2>&1).
I am running it from the cmd (DOS) prompt so I could remove the start..... I will try with just the >2

I was faster after the calibration but I do agree that this is getting me on the bike every day.. ;)

**Comment by Mickyduck55** [comment]
I will continue to use 3639 until the next step is here it has been reliable and stable and I will see if I can get some debug data but why the output to file not working is a mystery,(obviously something wrong with my syntax but according to the internet the syntax is correct).  The consul window always has the data displayed even though it should be switched off with the command line input.
Thanks and in case you missed it @mattipee rolling cal  tick tick tick

**Comment by Mickyduck55** [comment]
3639 45 min no issues..

**Comment by michelrdagenais** [comment]
You can try 3639 (debug + simspeed) without debugging (i.e. no --debug option) to see if the slowdown from debugging output makes the difference (with debug output it works) versus 3627 (simspeed no debug does not work well).

**Comment by Mickyduck55** [comment]
Just tried the "official" snapshot build which has some of the recent changes.. Motor brake info and rolling calibration.. sadly it dropped the connection at 17 Mins so looks like it has nothing to do with the changes...  When I go  back into train without exiting it asks if I have the brake switched on.  Closing GC and restarting the same message so I have to unplug the USB and reinsert to connect.

**Comment by Mickyduck55** [comment]
@michelrdagenais @mattipee 
3639 today with debug 35 mins no issues, without Debug 35 mins no issues.....

**Comment by mattipee** [comment]
Thanks, Micky. Sorry I've been quiet, I'll get back to GC soon.

Very odd, all this.

Do you still run Eric's VG artefact? Ever any issue with that?

**Comment by Mickyduck55** [comment]
@mattipee No problem.  Its a pity there is nothing I can run outside GC to monitor the port.  No not been using Erics build I can only cycle so far in a day LOL.
I have been running 3639 daily trying to get it to fall over but so far it has not.  I guess you read my previous note.. the snapshot build on the main site crashed.

**Comment by Mickyduck55** [comment]
3693 with debug today over an hour.. File attached I don't know if this is relevant as it overwrites in the consul window.  I had no noticeable issues while riding
[test.txt](https://github.com/GoldenCheetah/GoldenCheetah/files/5975876/test.txt)

**Comment by Mickyduck55** [comment]
@michelrdagenais  @mattipee 

[test.txt](https://github.com/GoldenCheetah/GoldenCheetah/files/5977676/test.txt)

No idea if this helps .....
Today 3637 usual error after about 10 mins.  I had Device monitoring studio running files attached that shows the disconnect

**Comment by Mickyduck55** [comment]
Hi Gents
Today using a different head unit.. Took Susans off her iMagic.  The only thing I cannot change is the brake and power pack as I don't have another
Same old story here.. tried 3637 bombed out almost immediately with "cannot connect to Fortius error".
3693 no problems 30 mins, so quit GC, loaded 3637 bombed out again... log attached 
[DEBUGTEXT.txt](https://github.com/GoldenCheetah/GoldenCheetah/files/5981378/DEBUGTEXT.txt)

**Comment by michelrdagenais** [comment]
In the latest test, DEBUGTEXT.txt, we see the "open" working, then every other communication with USB fails. Under those conditions, I don't see how it obtains the firmware version and other info that requires USB communications. Other than that, it seems that 3637 sometimes has problems, irrespective of the head unit, but not always. It looks like 3639 does not have problems but its timing is modified by the debug output.

I will propose a PR to improve debugging in master: options to send the output to a file and add a timestamp, and one to insert the debugging in Fortius. Then, you can track master, or branches based on master, and have more support for debugging when a problem happens.

**Comment by Mickyduck55** [comment]
Hi Michael
thank you for persisting with this ;)
I think your suggestion for a PR to address the debug is a good idea and
will be of use to the much wider GC community.  As I mentioned a few days
ago I had the same issue running the "Official" snapshot build from 5 Feb
downloaded from the main GC site so this is almost certainly not directly
related to the changes @mattipee <tunabiscuits796@gmail.com> has carried
out as they are not merged as yet ?
Is there nothing to be gained from using the sniffer software, if not Iwill
uninstall it.

**Comment by michelrdagenais** [comment]
@Mickyduck55 I looked at the code to produce debugging output. It behaves differently on Linux and Windows. How does it work for you? All debugging output is sent to the console?

On Linux, the output goes by default to stderr for debug builds, and to goldencheetah.log otherwise. The behavior is reversed with the --debug option (to .goldencheetah.log for debug builds and to stderr otherwise).

On Windows, it looks like this option does not do anything. You cannot get the output to goldencheetah.log. Are you using the --debug option, does it have any effect?

**Comment by Mickyduck55** [comment]
Hi it just goes to the console and seems to overwrite .. the file is only a
few hundred lines long and the initial data displayed as in brake and route
information is not in the file when I stop
I have tried - and — seems to make no difference and Ale did say it just
goes to console window and we are not able to redirect the data

**Comment by amtriathlon** [comment]
> On Windows, it looks like this option does not do anything. You cannot get the output to goldencheetah.log. Are you using the --debug option, does it have any effect?

Windows behavior is different in many aspects, to begin with GUI programs are different from console programs and don’t have access to a console unless one is created (GC can be built this way but users don’t like this), in https://github.com/GoldenCheetah/GoldenCheetah/issues/3481 I added the ability to attach to the parent console for debug output but this only partially emulates posix behavior and redirection is not supported .

Log to goldencheetah.log when not in debug mode is possible, I implemented this in https://github.com/GoldenCheetah/GoldenCheetah/pull/3485 but problems arose on Windows 8, I haven’t got that version for testing, it was not that relevant to me and reverted it. It works on Windows 10 if you want to use in a PR.

We have a logger for ANT+ protocol: https://github.com/GoldenCheetah/GoldenCheetah/blob/d50e1a8552b98a4e2545dbed3c7b9b08a4ac3de1/src/ANT/ANTLogger.h, perhaps something like that would be better if you want to log all protocol interactions.

**Comment by michelrdagenais** [comment]
> Log to goldencheetah.log when not in debug mode is possible, I implemented this in #3485 but problems arose on Windows 8, I haven’t got that version for testing, it was not that relevant to me and reverted it. It works on Windows 10 if you want to use in a PR.

Yes, this is what I have in mind, using installMessageHandler. With this, the behavior should be the same on all platforms. There are very few fprintf to stderr in GoldenCheetah. In a separate PR they could be converted to qDebug or qWarning, except perhaps for the output of --help / --usage.

We could have also options to specify the output file and the filter rules. When testing different cases, it is useful to output logging to different files. Moreover, the ~/.goldencheetah/goldencheetah.log file is not that obvious to find for users. Similarly, it is easier to tell users, when testing features, to add a command line option with logging rules, rather than tell them to add and edit a ~/.config/QtProject/qtlogging.ini file or set an environment variable. This is trivial to implement.

> We have a logger for ANT+ protocol: https://github.com/GoldenCheetah/GoldenCheetah/blob/d50e1a8552b98a4e2545dbed3c7b9b08a4ac3de1/src/ANT/ANTLogger.h, perhaps something like that would be better if you want to log all protocol interactions.

Yes, logging at the USB level would be a good place.

**Comment by Mickyduck55** [comment]
Been away for a couple of days so no riding.  Today 45 mins on 3693 without issue no debug running.  Im not sure that this build has ever disconnected.

**Comment by Mickyduck55** [comment]
Working away from home again.. so first ride in 3 days still no issues with 3693 1 hour .. and a bit ;)

**Comment by amtriathlon** [comment]
I think it would be useful to have a working baseline on master branch, so I suggest to try a version including only the functional Fortius changes without the low level protocol handling changes s.t. https://ci.appveyor.com/project/Joern-R/goldencheetah-knhd8/builds/37375815

**Comment by amtriathlon** [comment]
Since this doesn't seem to be converging, honestly I don't even know what 3693 is at this point nor why it is so relevant, perhaps we should go back to stablish working point in master branch:

> > 3609 no issues 31 mins
> 
> If you had the appetite, 3610 (msleep move) then 3620 (double reads) may help pin it down.

Which is the status of ci3610 and ci3620, anyone of these is stable?

**Comment by Mickyduck55** [comment]
Hi Ale
I will run 3610 and 3620 today and see how it goes although I do not think
these contain the recent debug added by @Michel Dagenais
<michelrdagenais@gmail.com>   The significance of 3693 for me is that so
far it has never dropped the connection.

**Comment by michelrdagenais** [comment]
> Log to goldencheetah.log when not in debug mode is possible, I implemented this in #3485 but problems arose on Windows 8, I haven’t got that version for testing, it was not that relevant to me and reverted it. It works on Windows 10 if you want to use in a PR.

I searched quite a bit and believe that I have a good solution in PR #3837. I have no way of testing it on Windows though. @Mickyduck55 could you test this artefact to see if you can get a log on Windows with the "--debug-file filename" option? You can list the options with --help. The extensive logging for Fortius is not there yet, this will be a subsequent PR.

> We have a logger for ANT+ protocol: https://github.com/GoldenCheetah/GoldenCheetah/blob/d50e1a8552b98a4e2545dbed3c7b9b08a4ac3de1/src/ANT/ANTLogger.h, perhaps something like that would be better if you want to log all protocol interactions.

Logging at the USB level is indeed a good idea. I looked at the ANTLogger and it seems that it logs detailed interactions all the time, there is no way to disable it. I could, instead of defining a new USBLogger class, simply use the facilities in Qt to get something similar, but simpler and more flexible. I coud define a "logging category" "goldencheetah.usb.debug" that would be disabled by default. When needing detailed logging, we could tell users to use the new --debug-rules option to enable it. The output would go in the same log file as the rest, but you could easily output the category name, and grep on that, to view separately the different log messages categories, if needed. What do you think?

**Comment by Mickyduck55** [comment]
Hi I will be home tomorrow so will test over the weekend
Thanks
Mick

**Comment by Mickyduck55** [comment]
Hi @michelrdagenais tried this today using the syntax:

START "" "C:\program files\golden cheetah\goldencheetah.exe" --debug -file C:\DEBUG\test1.txt

The file was created but is blank

--help displayed a range of options

![image](https://user-images.githubusercontent.com/22476749/110134805-b5611580-7dc5-11eb-811a-24def3563455.png)

**Comment by amtriathlon** [comment]
> Hi Ale I will run 3610 and 3620 today and see how it goes although I do not think these contain the recent debug added by @michel Dagenais <michelrdagenais@gmail.com> The significance of 3693 for me is that so far it has never dropped the connection.

Any news about 3610 or 3620?

**Comment by michelrdagenais** [comment]
> START "" "C:\program files\golden cheetah\goldencheetah.exe" --debug -file C:\DEBUG\test1.txt
> 
> The file was created but is blank

@Mickyduck55 it seems that you have a blank between --debug and -file, this should be a single option name without a space in between: --debug-file. Please retry without the space. Thanks!

**Comment by Mickyduck55** [comment]
Tried that earlier.. it just hangs at the screen below with the command
prompt cursor flashing
[image: image.png]

**Comment by Mickyduck55** [comment]
Hi @michelrdagenais  back home for the next few days, was there something wrong with my syntax on the command line ?

**Comment by amtriathlon** [comment]
> > Hi Ale I will run 3610 and 3620 today and see how it goes although I do not think these contain the recent debug added by @michel Dagenais [michelrdagenais@gmail.com](mailto:michelrdagenais@gmail.com) The significance of 3693 for me is that so far it has never dropped the connection.
> 
> Any news about 3610 or 3620?

The importance to test 3610/3620 was due to they introduced modifications to the Fortius protocol interactions, from @mattipee:

> > 3609 no issues 31 mins
> 
> If you had the appetite, 3610 (msleep move) then 3620 (double reads) may help pin it down.

I you are not willing to test those versions, current master has logging facilities @michelrdagenais  introduced to debug issues like this.

**Comment by Mickyduck55** [comment]
Thank you for the update I have been in touch with Michael and noted that
the usb debug had been added to the master... I already have it downloaded
so will use it when I next ride indoors

**Comment by amtriathlon** [comment]
Ok, then let's close this.
